Add experimental `useRequestContext` (#1471)
Shopify/hydrogen: docs/hooks/framework/userequestcontext.md

---
gid: 3fd2294b-4172-4185-81c9-501a41c6bd6c
title: useRequestContext
description: The useRequestContext hook provides access to a request-scoped context.
---

<aside class="note beta">
<h4>Experimental feature</h4>

<p>Request context is an experimental feature. As a result, functionality is subject to change. You can provide feedback on this feature by [submitting an issue in GitHub](https://github.com/Shopify/hydrogen/issues).</p>

</aside>

> The `useRequestContext` hook provides access to a request-scoped context to share data across multiple React Server Components while running on the server. It is useful for writing plugins and third-party libraries.

Reviewer: ```suggestion
The `useRequestContext` hook provides access to a request-scoped context to share data across multiple React Server Components while running on the server. It's useful for writing plugins and third-party libraries.
```

## Example code

{% codeblock file, filename: 'App.server.jsx' %}

```jsx
import {useRequestContext} from '@shopify/hydrogen';

const pluginName = 'my-plugin';

function installMyPlugin() {
  const context = useRequestContext(pluginName);
  context.installed = true; 
}

function useMyPlugin() {
  return useRequestContext(pluginName);
}

function App() {
  installMyPlugin();
  
  return /* ... */;
}

function SomeOtherServerComponent() {
  const data = useMyPlugin(); 
  return <div>{data.installed}</div>
}
```

{% endcodeblock %}

## Arguments

The `useRequestContext` hook takes the following arguments:

| Key     | Required | Description                                                              |
| ------- | -------- | ------------------------------------------------------------------------ |
| `scope` | No       | A string to encapsulate data so that it's not modified by other scopes.  |

## Return value

The `useRequestContext` hook returns an object that can be mutated to store data.

## Considerations

> - Consider using the `useRequestContext` hook only where appropriate. Generally, you should pass props down to the components instead.

Author: Alright, I've added [this](https://github.com/Shopify/hydrogen/pull/1471/commits/ef17705a904566ebd47aff14893c7b8af91d0401).

> - Consider using the `useRequestContext` hook only where appropriate. Generally, you should pass props down to the components instead.

Reviewer: Perfect! Thanks, @frandiox!

> - The `useRequestContext` hook is only available when running on the server (Server Components, or Client Components during SSR), and never shared in the browser.

Reviewer: ```suggestion
- The `useRequestContext` hook is only available when running on the server in server components or client components during server-side rendering (SSR). The hook is never shared in the browser.
```
